### DISCUSSION TITLE
Why does changing the type of `_id` change the return type of `populate`?

### DISCUSSION THREAD

**Discussion by kiancross** [typegoose/typegoose#481] [Q&A] [answered]
I have changed the `_id` type (described [here](https://typegoose.github.io/typegoose/docs/guides/advanced/change-id-type)) of `Forum` to `string`:

```ts
import { Ref, prop, getModelForClass } from "@typegoose/typegoose";

export class Message {
  @prop({ required: true, ref: () => Forum })
  public parent: Ref<Forum>;
}

export class Forum {
  @prop()
  public _id!: string;

  @prop({
    ref: () => Message,
    foreignField: "parent",
    localField: "_id",
  })
  public messages?: Ref<Message[]>;
}

export const MessageModel = getModelForClass(Message);
export const ForumModel = getModelForClass(Forum);
```

However this changes the return type of `foundForum.populate` from `DocumentType<Forum>` to `mongoose.Document`.

```ts
const forum = await ForumModel.create({_id: "foo"});

await MessageModel.create({parent: forum._id});

const foundForum = await ForumModel.findById(forum.id);
console.log(foundForum);

if (foundForum !== null) {
  const populated = await foundForum.populate("messages").execPopulate();
  console.log(populated);
}
```

Why is this the case?

Thanks a lot :)

**Comment by hasezoey** [comment, marked as answer]
firstly, your array type is wrong, its `Ref<Class>[]` not `Ref<Class[]>`

aside from that, in your `Message` class you didnt define the `_id` type of `Forum`, does it change anything when you use:
```ts
@prop({ required: true, ref: () => Forum, type: String })
public parent: Ref<Forum, string>;
```

PS: are you sure the type is just `mongoose.Document`? when not, try running it through [`isDocumentArray`](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-document#isdocumentarray)

PPS: what is your mongoose version? highest currently supported is `5.10.18` (#432)

**Reply by kiancross** [reply]
> firstly, your array type is wrong, its `Ref<Class>[]` not `Ref<Class[]>`

Thanks - fixed this.

> aside from that, in your `Message` class you didnt define the `_id` type of `Forum`, does it change anything when you use:

Fixed that, but it didn't change anything.

> PS: are you sure the type is just `mongoose.Document`?

Yes - if the custom `_id` is removed, the type is `DocumentType<Forum>`. When the custom `_id` type is added, the type is `mongoose.Document`.

> when not, try running it through [`isDocumentArray`](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-document#isdocumentarray)

It doesn't compile:
```
Property 'messages' does not exist on type 'Document'.
```

> PPS: what is your mongoose version? highest currently supported is `5.10.18` (#432)

Using version `5.10.18`.

**Reply by kiancross** [reply]
I've put an example repo on GitHub, if that helps: https://github.com/kiancross/typegoose-test.

**Reply by hasezoey** [reply]
@kiancross thanks for this repo, i didnt have time to look into this before - but this speeds it up

---

from what i can tell, there is nothing typegoose can do about that type problem [at least with the current project's type structures] (of the return type from `populate` or `execPopulate`), but from my knowledge, you dont needs the return value, because it modifies the input, so the input can be used instead (which has the correct type)

```ts
const doc = await Model.findOne();
const pop = await doc.populate().execPopulate();
pop === doc;
```
Note: this above is "pseudo code" and the equals my return false, but this is meant symbolicly

**Reply by kiancross** [reply]
Thanks for the help - that makes sense - I'll use the above.
